lifecycle/lifecycle-runtime/src/nativeTest/kotlin/androidx/lifecycle/runLifecycleTest.native.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.lifecycle
+
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.native.concurrent.ThreadLocal
+import kotlinx.coroutines.CloseableCoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.MainCoroutineDispatcher
+import kotlinx.coroutines.Runnable
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.TestResult
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+
+@ThreadLocal
+private var isMainThread = false
+
+/**
+ * On native platforms, tests are running from the main thread itself,
+ * so it is impossible to reschedule some coroutines to a later time,
+ * as the test code itself is preventing it.
+ *
+ * That's why we need a complete replacement for the Dispatchers.Main.
+ */
+@OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
+private class SurrogateMainCoroutineDispatcher : MainCoroutineDispatcher() {
```

Comment:
Can we remove duplication with the other file, moving it to commonTest?

Reply:
ThreadLocal is a class on JVM/Android, but an annotation on a global (!) variable in Native, which makes it quite difficult to share them on different platforms. From my point of view, attempting to work around these differences will make the code more complicated and less readable.

Reply:
Okay